### ISSUE TITLE
More intelligent file scanner

### ISSUE THREAD

**Issue by PaulWoitaschek** [PaulWoitaschek/MaterialAudiobookPlayer#299]
Write a more intelligent file scanner who is able to scan only once on startup.
After that watch all directories recursively and update / add / remove the books accordingly.

**Comment by j-opificius** [comment]
Yes! Player doesn't notice when I add books to my book directory. Maybe a "rescan" function ?

**Comment by Ph1b** [comment]
It should recognize them as of today. Maybe you added them in a wrong folders, or the files are defect, or its some special file format?

**Comment by j-opificius** [comment]
I'll try again with the latest version. It was at least partly my 
misunderstanding of how collective folders work: I had two book 
file-sets in the same folder, and it appended the second set onto the 
first, so I didn't see the second book.
Again, I'll get a better understanding of how the library mechanism 
works before I squawk again :-)

Julian.

**Comment by Ph1b** [comment]
Well if you misunderstood it it means that it is not clear enough explained. Do you have a suggestion to prevent these misunderstandings?

**Comment by j-opificius** [comment]
That's very modest of you, Paul :-)

Challenge accepted; let me get properly familiar with it and I would be 
more than happy to propose a description for you to consider. Proper 
handling of audiobooks is important to me, and I appreciate your 
commitment to providing an application tuned to that purpose. I would be 
very happy to contribute some effort to the documentation.

Julian.

**Comment by j-opificius** [comment]
Paul,

I'm having trouble with basics now, and my installation is essentially 
broken. The app crashes when I navigate to the track selector (touching 
on the current track indicator), so I can't even select tracks.

I'm having trouble rebuilding my folder now. Do you have a document 
somewhere that describes the required folder layout and file naming 
structure?

Thanks,
Julian Opificius.

**Comment by Ph1b** [comment]
Which version are you using? If you find a bug please report on a separate issue.

The file structure is not very complicated.

Either you add a single book which means all contents inside that folder (including subfolders) will be recognized as a a single book. (i.e /storage/.../HarryPotter1/

Or you have a multiple folder one like /storage/../Audiobooks/
That is the same as adding each folder inside Audiobooks as a single book, but it will automatically add new ones.

**Comment by j-opificius** [comment]
I understand the individual _words_, but not the _meaning_ of what you 
say, and your description is not complete. _You_ understand what you're 
doing, but you don't describe it completely for your users :-)

How does the app resolve a book title name? For example, do you use the 
"<disk title>.m3u8" file created by my ripping program (I use Exact 
Audio Copy - "EAC") ? Or is there some other method?

How does the app expect to see individual chapter files named ? 
"<disk-num>-<track-num>.mp3" for example?

What filename structure does the app want to see for files from each 
disk of a single book with multiple disks?

I have a folder structure as follows:
/storage/Author1/Title1 - disk 01/<disk-num>-<track-num>.mp3
/storage/Author1/Title1 - disk 02/<disk-num>-<track-num>.mp3
...
/storage/Author1/Title2 - disk 01/<disk-num>-<track-num>.mp3
/storage/Author1/Title2 - disk 02/<disk-num>-<track-num>.mp3
...
/storage/Author2/Title1 - disk 01/<disk-num>-<track-num>.mp3
/storage/Author2/Title1 - disk 02/<disk-num>-<track-num>.mp3
...

This has worked for me for months, except that the title of the book has 
been shown as "Title1 - disk 01", which is not really the best.

Today, after update, it crashes continuously. Sometimes I have to 
completely delete thee folder structure to get it working again.

Maybe this is a clue:
As of your latest change, the disk/track indicator indicates incorrect 
information:
1 - -01
2 - 01-02
3 - 01-03
...

This did not happen previously - say, before Friday,
And the app crashes continuously when I try to navigate.

Where is the  version shown? I don't see it anywhere.

Where are configuration/library files stored?

Julian.

**Comment by irregularisnotrandom** [comment]
Why not offer a way to do manual scan only?

**Comment by PaulWoitaschek** [comment]
For convenience.

**Comment by irregularisnotrandom** [comment]
I personally think that a manual scan is much more convenient solution, given that i can exrend my battery hold that way.

**Comment by PaulWoitaschek** [comment]
Did you measure the duration?

**Comment by irregularisnotrandom** [comment]
I did not but i see that couple minutes min eachtime it rescans. i am sure thus number will go up as I add new audio. also the scan is not one time you kniw. It foes this multiple times during the life of the app usage.

**Comment by irregularisnotrandom** [comment]
Sorry for the typos I am on a phone and my addons block rediting scripts

**Comment by PaulWoitaschek** [comment]
I don't expect that it costs too much battery.

Most of the times users stay in a book and the scanner only gets triggered when you enter the book shelf.

**Comment by irregularisnotrandom** [comment]
Actually that logic does not hold well and the reason is this. What is the likely hood of recurring adding multiple new books to the lib by the user while the app is open or closing? No user is going to add new audio book every 10-20 mins to the lib everytime they use the app.

**Comment by PaulWoitaschek** [comment]
That is the reason why this ticket exists.

**Comment by irregularisnotrandom** [comment]
To me this auto scan is just causing unnecessary energy use. at lwast that is how i personally see theis issue. Manual scan is not unheard of, many apps take that route.

**Comment by PaulWoitaschek** [comment]
How do you actually know this? What do the battery stats show?

**Comment by irregularisnotrandom** [comment]
Well any disk activity will cause energy consumption and this app does that activity a lot. but in all fairness i will try to find and up to stat it to be precise.

**Comment by bitinerant** [comment]
I would love a manual scan option--not so much for battery life in my case, but because when I add books, it seems to take forever for them to show up in MP.  For some reason, closing and re-opening the app hasn't helped me in the past--I have to just sit and wait for the books to appear in the list.  (I haven't tested, but one ebook with 1218 files may be causing the delay.)

**Comment by sandybeachit** [comment]
There are two issues with the current scanner on android 6 from what i can tell #460  and #475 

There seems to be a bug where the app will only display the first book of a series listed in a subfolder such as: Audtiobooks > Author > Series Name > Book 1, 2 etc

Take a look at the above two issues #460 and #475 and your see the current bug, perhaps this could be fixed while a rewrite of the scanner is happening ?

Thanks

David

**Comment by PaulWoitaschek** [comment]
An intelligent file scanner is not necessary as the current one works perfectly.

However I just introduced a change that massively increase the scanning time.

I added a timestamp to each chapter that contains the time of the last modification of the file. When the file to scan has the same timestamp, that chapter will be used and the metadata won't be scanned at all.
For my music collection (about 16 gigs) it decreased the scanning time from 1m:30s to about 0.5 seconds.

The first scan will however take its time, but there is no way around that.

Can you test the version and tell me if it works as expected?

https://github.com/PaulWoitaschek/MaterialAudiobookPlayer/releases/tag/v3.0.13-RC

**Comment by PaulWoitaschek** [comment]
You can also grab the latest version by joining the beta:
https://play.google.com/apps/testing/de.ph1b.audiobook
